src/Services/PrepareService.php

```diff
@@ -80,28 +82,47 @@ public function transferLazy(
         Wallet $from,
         Wallet $to,
         string $status,
-        float|int|string $amount,
+        CostDtoInterface|float|int|string $amount,
         ExtraDtoInterface|array|null $meta = null
     ): TransferLazyDtoInterface {
         $discount = $this->personalDiscountService->getDiscount($from, $to);
         $toWallet = $this->castService->getWallet($to);
+        $cost = $this->castService->getCost($amount);
         $from = $this->castService->getWallet($from);
-        $fee = $this->taxService->getFee($to, $amount);
+        $fee = $this->taxService->getFee($to, $cost->getValue());
 
-        $amountWithoutDiscount = $this->mathService->sub($amount, $discount, $toWallet->decimal_places);
+        $amountWithoutDiscount = $this->mathService->sub($cost->getValue(), $discount, $toWallet->decimal_places);
         $depositAmount = $this->mathService->compare($amountWithoutDiscount, 0) === -1 ? '0' : $amountWithoutDiscount;
         $withdrawAmount = $this->mathService->add($depositAmount, $fee, $from->decimal_places);
+        $currencyRate = $cost->getCurrency() === null
+            ? 1
+            : $this->exchangeService->convertTo($cost->getCurrency(), $from->currency, 1);
+
+        $withdrawCost = $this->mathService->mul($currencyRate ?? 1, $withdrawAmount);
+        $withdrawCostFloor = $this->mathService->floor($withdrawCost);
+        $depositCost = $this->mathService->mul($currencyRate ?? 1, $depositAmount);
+        $depositCostFloor = $this->mathService->floor($depositCost);
         $extra = $this->extraDtoAssembler->create($meta);
         $withdrawOption = $extra->getWithdrawOption();
         $depositOption = $extra->getDepositOption();
+        $depositMeta = $cost->getCurrency() === null
+            ? $depositOption->getMeta()
+            : array_merge(
+                $depositOption->getMeta() ?? [],
+                [
+                    'cost' => $depositCostFloor,
+                    'currency' => $cost->getCurrency(),
+                ]
```

Comment:
I don't like the metadata limitation. This has always been freedom, but when setting up internal parameters, we will be forced to describe the limitations in the documentation.